### DISCUSSION TITLE
Using Dirac kernels with reporters

### DISCUSSION THREAD

**Discussion by salaudeen-ya** [idaholab/moose#24437] [Q&A General] [answered]
Hello,

I am using a method of perfectly matched discrete layers to solve a simple problem in 2D. Note that I have successfully used the same technique to solve 1D problems and it works fine. But I have huge inconsistent results with the 2D case. The results are blowing up and not converging as expected.

I have the following results: The sampled variable at the interface before the iteration is u = 3.92415 and 3.88704, xflux = 0.196009 and 0.19600, v = 0 and 0, xflux_v = 0 and 0. After the iteration, u = 3.92415 and 3.88704, xflux = 0.196009 and 0.196009, xflux_v = 0.72031 and 0.72031, v = 67.38297 and 67.37507, xflux_v = 0.72031 and 0.72031. 

These results are severely not correct and they are obtained from the following files:

First subsystem:
```
[Mesh]
  [./generated]
    type = CartesianMeshGenerator
    dim = 2
    dx = '1 1'
    ix = '32 1'
    dy = '0.1'
    iy = '1'
    subdomain_id = '1 2'
  [../]
[]


[Variables]
  [u]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
  [xflux]
    order = CONSTANT
    family = MONOMIAL
  []
  [yflux]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[Kernels]
  [diffusion]
    type = MatDiffusion
    variable = u
    diffusivity = 1
  []
    [reaction]
    type =  Reaction
    variable = u
    rate = 0.25
  []
[]

[AuxKernels]
  [xfluxKernel]
    type = DiffusionFluxAux
    diffusivity = -1
    variable = xflux
    diffusion_variable = u
    component = x
  []
  [yfluxKernel]
    type = DiffusionFluxAux
    diffusivity = -1
    variable = yflux
    diffusion_variable = u
    component = y
  []

[]

[DiracKernels]
  [point_source]
    type = ConstantPointSource
    variable = u
    value = 1
    point = '0.8 0.0 0'
    point_not_found_behavior = 'ERROR'
  []

  
  [unode1_interface]
    type = ReporterPointSource
    variable = u
    value_name = 'csv_v_interface/v'
    x_coord_name = 'csv_v_interface/x'
    y_coord_name = 'csv_v_interface/y'
    z_coord_name = 'csv_v_interface/z'
    #point_name = 'csv_v_interface/x'
    point_not_found_behavior = 'ERROR'
  []
  [xfluxnode1_interface]
    type = ReporterPointSource
    variable = u
    value_name = 'csv_flux_interface/xflux_v'
    x_coord_name = 'csv_flux_interface/x'
    y_coord_name = 'csv_flux_interface/y'
    z_coord_name = 'csv_flux_interface/z'
    point_not_found_behavior = 'ERROR'
  []

[]

[BCs]
  [leftBC] 
    type = DirichletBC
    variable = u
    boundary = 'left'
    value = 0
  []
  [rightBC] 
    type = DirichletBC
    variable = u
    boundary = 'right'
    value = 0
  []
[]

[VectorPostprocessors]
  [u_interface]
    type = LineValueSampler
    variable = 'u'
    start_point = '0.999999 0 0'
    end_point = '0.999999 0.099999 0'
    scaling =   1.0625 #0.033203125   
    num_points = 2
    sort_by = x
  []
  [flux_interface]
    type = LineValueSampler
    variable = 'xflux'
    start_point = '0.999999 0 0'
    end_point = '0.999999 0.099999 0'
    scaling = -0.05  
    num_points = 2
    sort_by = x
  []
  [csv_v_interface]
    type = CSVReader
    csv_file = left_subdomain_pmdl_schwarz_new_out_sub0_v_interface_0001.csv
  []
  [csv_flux_interface]
    type = CSVReader
    csv_file = left_subdomain_pmdl_schwarz_new_out_sub0_flux_interface_0001.csv
  []
[]


[Executioner]
  type = Steady
  solve_type = 'NEWTON'

  fixed_point_max_its = 1000
  nl_abs_tol = 1e-10
  fixed_point_rel_tol = 1e-16
  fixed_point_abs_tol = 1e-10

  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'

  [Quadrature]
    type = 'GAUSS'
    custom_blocks = '1 2'    
    custom_orderX = 'FIRST CONSTANT'
    custom_orderY = 'FIRST FIRST'
  []

[]


[Outputs]
  exodus = true
  csv = true
  console = true
[]


[MultiApps]
  [sub]
    type = FullSolveMultiApp
    #positions = '0.5 0 0'
    input_files = 'right_subdomain_pmdl_schwarz_new.i'
    execute_on = 'timestep_begin'
  []
[]

```

Second subsystem:
```
[Mesh]
  [./generated]
    type = CartesianMeshGenerator
    dim = 2
    dx = '1 1'
    ix = '1 32'
    dy = '0.1'
    iy = '1'
    subdomain_id = '1 2'
  [../]
[]

[Variables]
  [v]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
  [xflux_v]
    order = CONSTANT
    family = MONOMIAL
  []
  [yflux_v]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[Kernels]
  [diffusion]
    type = MatDiffusion
    variable = v
    diffusivity = 1
  []
  [reaction]
    type =  Reaction
    variable = v
    rate = 0.25
  []
[]

[AuxKernels]
  [xfluxKernel]
    type = DiffusionFluxAux
    diffusivity = -1
    variable = xflux_v
    diffusion_variable = v
    component = x
  []
  [yfluxKernel]
    type = DiffusionFluxAux
    diffusivity = -1
    variable = yflux_v
    diffusion_variable = v
    component = y
  []
[]

[DiracKernels]
  [vnode1_interface]
    type = ReporterPointSource
    variable = v
    value_name = 'csv_u_interface/u'
    x_coord_name = 'csv_u_interface/x'
    y_coord_name = 'csv_u_interface/y'
    z_coord_name = 'csv_u_interface/z'
    #point_name = 'csv_v_interface/x'
    point_not_found_behavior = 'ERROR'
  []
  [xfluxnode1_interface]
    type = ReporterPointSource
    variable = v
    value_name = 'csv_flux_interface/xflux'
    x_coord_name = 'csv_flux_interface/x'
    y_coord_name = 'csv_flux_interface/y'
    z_coord_name = 'csv_flux_interface/z'
    #point_name = 'csv_v_interface/x'
    point_not_found_behavior = 'ERROR'
  []
[]

[BCs]

  [leftBC] 
    type = DirichletBC
    variable = v
    boundary = 'left'
    value = 0
  []


[]

[VectorPostprocessors]
  [v_interface]
    type = LineValueSampler
    variable = 'v'
    start_point = '1.000001 0 0'
    end_point = '1.000001 0.099999 0'
    scaling =  1.0625 #0.033203125    
    num_points = 2
    sort_by = x
  []
  [flux_interface]
    type = LineValueSampler
    variable = 'xflux_v'
    start_point = '1.000001 0 0'
    end_point = '1.000001 0.099999 0'
    scaling = 0.05  
    num_points = 2
    sort_by = x
  []
  [csv_u_interface]
    type = CSVReader
    csv_file = left_subdomain_pmdl_schwarz_new_out_u_interface_0001.csv
  []
  [csv_flux_interface]
    type = CSVReader
    csv_file = left_subdomain_pmdl_schwarz_new_out_flux_interface_0001.csv
  []
[]


[Executioner]
  type = Steady
  solve_type = 'NEWTON'



  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
  
  [Quadrature]
    type = 'GAUSS'
    custom_blocks = '1 2'    
    custom_orderX = 'CONSTANT FIRST'
    custom_orderY = 'FIRST FIRST'
  []
[]



[Outputs]
  exodus = true
  csv = true
  console = true
[]
```

Note that I have a locally modified libmesh that applied quadrature in different directions as I need the midpoint rule in the domain with a unit length discretization. 

I will appreciate why the results obtained from these files are not consistent.
A simple 1D of the problem is given as follows and I have used the same PMDL technique to solve it without using the reporter since I can simply manually post-process the result to transfer without writing it in a file:
```
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 1000
  xmin = 0
  xmax = 2
[]

[Variables]
  [u]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
  [flux]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[Kernels]
  [diffusion]
    type = MatDiffusion
    variable = u
    diffusivity = 1
  []
  [reaction]
    type =  Reaction
    variable = u
    rate = 0.25
  []
[]

[DiracKernels]
  [point_source]
    type = ConstantPointSource
    variable = u
    value = 1.0
    point = '0.8 0 0'
  []
[]

[AuxKernels]
  [fluxKernel]
    type = DiffusionFluxAux
    diffusivity = -1
    variable = flux
    diffusion_variable = u
    component = x
  []
[]

[BCs]
 [leftBC] 
    type = DirichletBC
    variable = u
    boundary = 'left'
    value = 0
  []
  
[]



[Executioner]

  type = Steady
  solve_type = 'PJFNK'

  fixed_point_max_its = 500
  nl_abs_tol = 1e-10
  fixed_point_rel_tol = 1e-16
  fixed_point_abs_tol = 1e-10

  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  exodus = true
  #csv = true
  console = true
[]
```
Thank you !!

**Comment by GiudGiud** [comment]
Hello

Does this work fine with a regular quadrature?
Basically, are the results wrong because of the new quadrature you made? Or is this something else?

Guillaume

**Reply by salaudeen-ya** [reply]
I have a similar issue without specifying the quadrature block. It should be something else.

**Reply by GiudGiud** [reply]
So Gauss is the default quadrature anyway, so specifying
```
  [Quadrature]
    type = 'GAUSS'
    custom_blocks = '1 2'    
    custom_orderX = 'CONSTANT FIRST'
    custom_orderY = 'FIRST FIRST'
  []
```
only worked to change the order. If you changed the Gauss quadrature, you would be using the changed one whether you specify it or not.

I would try a non-modified quadrature first. 

If you're confident the problems are not coming from your modified quadrature, can you specify what the issue is with the results?

**Reply by salaudeen-ya** [reply]
I have validated the implementation of the quadrature rule by using it on a rotated simple anisotropic 2D problem in different directions.

That being said, consider the results obtained from these files:

![first](https://github.com/idaholab/moose/assets/38023107/57c4db88-e547-4d9a-860e-3e15d7b6e423)
![second](https://github.com/idaholab/moose/assets/38023107/9c038200-e9ac-4d72-9a58-59547b930190)

You can notice the huge discrepancy between the results. At least, the result should be pretty close to each other and to the reference result

![single](https://github.com/idaholab/moose/assets/38023107/96018727-38cb-45d2-a9a8-36df3cf9edc8)


From the sampled values before the start of the coupling is u = 3.92415 and 3.88704, xflux = 0.196009 and 0.19600, v = 0 and 0, xflux_v = 0 and 0. After the iteration, both u and xflux values do not change while xflux_v = 0.72031 and 0.72031 and v = 67.38297 and 67.37507. My worry is justified since both u and xflux values do not change even after the simulation, but v has a drastic increase in value.

**Reply by GiudGiud** [reply]
Are these results converged? Like mesh-wise
What are u & v physically?
Can you decouple u & v, keeping one constant? To verify the equation for u and v separately

**Reply by salaudeen-ya** [reply]
Both u and v physically are basically the same variable in two different subdomains. 

So, I tried by basically a 2D version of a 1D problem with the force equally on the nodes, and comparing it with my 1D result which agrees.

However, that involves me manually postprocessing the result and this method adds several lines to the input files, even for one element (two nodes) in the y-direction.

That was why I tried using reporters to allow me to specify any number of discretizations.

**Reply by GiudGiud** [reply, marked as answer]
Reporter is a good idea to postprocess.

I would try to s simplify the problem until you get the expected result. This will help you narrow down where the issue here.

**Reply by salaudeen-ya** [reply]
I tried narrowing down my problem to a simple case of having one element with two nodes in the y-direction. Then I manually post-process the result as I want it, before using the same construct for the VectorPostprocessor, after which I could get several numbers of discretization in the y-direction.

Thank you.
